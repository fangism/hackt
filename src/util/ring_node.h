/**
	\file "ring_node.h"
	Declaration for ring_node struct.
	$Id: ring_node.h,v 1.1.2.2.2.1 2005/02/07 20:16:56 fang Exp $
 */

#ifndef	__UTIL_RING_NODE_H__
#define	__UTIL_RING_NODE_H__

// whether or not to include subset of definitions separately, but still inline
#ifndef	FORCE_INLINE_RING_NODE
#define	FORCE_INLINE_RING_NODE		0
#endif

// #include <iostream>
#include <iterator>
#include "macros.h"
#include "ring_node_fwd.h"

namespace util {
//=============================================================================
// forward declarations

class ring_node_base;

class ring_node_iterator_base;

template <class, class, class>
class ring_node_iterator;

//=============================================================================
/**
	Safely down-castable base class for templated ring_nodes.  
 */
class ring_node_base {
friend class ring_node_iterator_base;
protected:
	ring_node_base*				next;

private:
	/**
		All ring_node point to themselves upon construction.  
		No empty construction.  
	 */
	ring_node_base();

protected:
	explicit
	ring_node_base(ring_node_base* r) : next(r) { NEVER_NULL(next); }

#if FORCE_INLINE_RING_NODE
	/**
		Potentially expensive destructor.
		Must maintain circular reference invariant!
	 */
	~ring_node_base() {
		if (next != this) {
			ring_node_base* walk = next;
			while (walk->next != this)
				walk = walk->next;
			// found the node that points to this, update it
			walk->next = next;
		}
		// else this is the last node, just drops itself
	}
#else
	~ring_node_base();
#endif

	/**
		Fusing two ring_nodes into one, extremely efficient.  
		This is unsafe, because if two nodes are already
		in the same cycle, then swapping pointer will result
		in two disjoint cycles!  Kinda cool actually.  
	 */
	void
	unsafe_merge(ring_node_base& r) {
		std::swap(next, r.next);
	}

public:
	/**
		Checks for intersection before merging.  
		If they already intersect, then this does nothing.  
	 */
	void
	safe_merge(ring_node_base& r) {
		if (!contains(r))
			unsafe_merge(r);
	}

	/**
		Run-time check for intersection.  
		Stops as soon as one element is found reachable from the other, 
		or as soon as one list is done searching.
		\pre A containing B <=> B containing A.
		\post A containing B <=> B containing A.
	 */
#if FORCE_INLINE_RING_NODE
	bool
	contains(const ring_node_base& r) const {
		const ring_node_base* walk1 = this;
		const ring_node_base* walk2 = &r;
		do {
			if (walk1 == &r || walk2 == this)
				return true;
			else {
				walk1 = walk1->next;
				walk2 = walk2->next;
			}
		} while (walk1 != this && walk2 != &r);
		return false;
	}
#else
	bool
	contains(const ring_node_base& r) const;
#endif

	bool
	points_to(ring_node_base* b) const {
		return (next == b);
	}

};	// end struct ring_node_base

//-----------------------------------------------------------------------------
/**
	Base iterator for generic ring nodes.  
 */
class ring_node_iterator_base {
template <class, class, class> friend class ring_node_iterator;
public:
	typedef	size_t				size_type;
	typedef	ptrdiff_t			difference_type;
	typedef	std::forward_iterator_tag	iterator_category;
protected:
	/**
		Points to the current referenced node.
	 */
	ring_node_base*				current_node;
	/**
		Since iterator points to a circular structure with no
		one-past-the-end element, we need a second pointer to
		distinguish between between begin and end when iterators
		point to the same reference element.  
		See implementation of comparison operators.  
	 */
	const ring_node_base*			last_node;

protected:
	// protecting constructors to prevent abuse...

	ring_node_iterator_base() : current_node(NULL), last_node(NULL) { }

	/**
		This constructor is intended for use with begin().  
		Note that by construction, no ring_nodes are empty!
	 */
	explicit
	ring_node_iterator_base(ring_node_base* p) :
		current_node(p), last_node(NULL) {
	}

	/**
		This constructor is intended for use with end().
		\param p initial node pointer.  
		\param l should be equal to p->next, unless it is NULL.
	 */
	ring_node_iterator_base(ring_node_base* p, const ring_node_base* l) :
		current_node(p), last_node(l) {
		NEVER_NULL(l);
		INVARIANT(l->points_to(p));
	}

public:
	void
	advance(void) {
		last_node = current_node;
		current_node = current_node->next;
		NEVER_NULL(last_node);
		NEVER_NULL(current_node);
	}

	/**
		Most of the time, this should be as fast as 
		the time for one comparison because the first
		comparison will usually be short-circuit false.  
	 */
	bool
	operator == (const ring_node_iterator_base& i) const {
		return (current_node == i.current_node) &&
			(last_node == i.last_node);
	}

	/**
		Most of the time, this should be as fast as 
		the time for one comparison because the first
		comparison will usually be short-circuit true.  
	 */
	bool
	operator != (const ring_node_iterator_base& i) const {
		return (current_node != i.current_node) ||
			(last_node != i.last_node);
	}

	bool
	operator == (const ring_node_base* r) const {
		return current_node == r;
	}

	bool
	operator != (const ring_node_base* r) const {
		return current_node != r;
	}

#if 0
	std::ostream&
	dump(std::ostream& o) const {
		return o << "ring_node_iter: " << current_node <<
			", " << last_node;
	}
#endif

};	// end class ring_node_iterator_base

//-----------------------------------------------------------------------------
/**
	A ring_node::iterator.
 */
template <class T, class R, class P>
class ring_node_iterator : public ring_node_iterator_base {
public:
	typedef	T				value_type;
	typedef	R				reference;
	typedef	P				pointer;

	typedef	ring_node_iterator<T, R, P>	this_type;
	typedef	ring_node_iterator<T, T&, T*>	iterator;
	typedef	ring_node_iterator<T, const T&, const T*>
						const_iterator;
	typedef	ring_node<T>			node_type;
	typedef	ring_node_base			node_base_type;

	explicit
	ring_node_iterator(node_base_type* n) :
		ring_node_iterator_base(n) { }

	/**
		\pre n must point to l.
	 */
	ring_node_iterator(ring_node_base* n, const ring_node_base* l) :
		ring_node_iterator_base(n, l) { }

	// using default copy constructor
	// iterators are allowed to point to the same nodes.  

	/**
		Allows both iterator and const_iterator to be 
		copy-constructed from iterator.  
	 */
	// explicit
	ring_node_iterator(const iterator& i) :
		ring_node_iterator_base(i.current_node, i.last_node) { }

	reference
	operator * () const {
		return *static_cast<node_type&>(*current_node);
	}

	pointer
	operator -> () const { return &(operator * ()); }

	/**
		Prefix increment / advance.
	 */
	this_type&
	operator ++ () {
		this->advance();
		return *this;
	}

	/**
		Postfix increment / advance.
	 */
	this_type
	operator ++ (int) {
		const this_type temp(*this);
		this->advance();
		return temp;
	}

};	// end struct ring_node_iterator

//-----------------------------------------------------------------------------
/**
	A ring-node is an implementation of an amoeba-like circularly
	linked list.  
	A ring_node by itself forms a ring with itself.  
	Any two ring_node merged together will connect into a bigger ring.  

	Nodes of this type have several invariant properties:
	1) Nodes are self assembled into rings.  
		Any node object is a member of EXACTLY ONE ring.  
		A new node is always connected to itself in a self-ring.  
	2) As rings grow and nodes die, this property is maintained INVARIANT.  
	3) If A is reachable from B <=> B is reachable from A.  
	4) While shape and structure are maintained, nodes never manage
		each other's memory -- strictly mind-your-own-memory.  
 */
template <class T>
class ring_node : public ring_node_base {
template <class, class, class> friend class ring_node_iterator;
private:
	typedef	ring_node_base			base_type;
	typedef	ring_node<T>			this_type;
public:
	typedef	T				value_type;
	typedef	T&				reference;
	typedef	const T&			const_reference;
	typedef	T*				pointer;
	typedef	const T*			const_pointer;
	typedef	ring_node_iterator<T, T&, T*>	iterator;
	typedef	ring_node_iterator<T, const T&, const T*>
						const_iterator;
	// no reverse iterators, forward only
	typedef	size_t				size_type;
	typedef	ptrdiff_t			difference_type;
	// any allocator type
protected:
	value_type				value;
public:
	/**
		All ring_node point to themselves upon construction.  
	 */
	ring_node() : ring_node_base(this), value() { }

	explicit
	ring_node(const value_type& v) : ring_node_base(this), value(v) { }

	/**
		NOTE the unusual copy construct semantics!
		Copy constructor does not copy the ring!
		Only copies the value contained.  
		This maintains the uniqueness property:
		No ring_node shall be pointed to be more than one other
		ring_node.  
	 */
	ring_node(const ring_node& r) : ring_node_base(this), value(r.value) { }

	/**
		Uses base destructor, which walks the list.  
		Will not be virtual, never deleting through base-type 
		pointers.  
	 */
	~ring_node() { }

	// no assignment

	/**
		Since node and list are one and the same, 
		we can dereference the node just like an iterator.  
	 */
	reference
	operator * () { return value; }

	const_reference
	operator * () const { return value; }

	pointer
	operator -> () { return &value; }

	const_pointer
	operator -> () const { return &value; }

	iterator
	begin(void) { return iterator(next); }

	const_iterator
	begin(void) const {
		return const_iterator(next);
	}

	iterator
	end(void) { return iterator(this->next, this); }

	/**
		Nodes and lists are actually one and the same
		tightly coupled.  
	 */
	const_iterator
	end(void) const {
		return const_iterator(this->next, this);
	}

	/**
		ring_nodes are never empty!
	 */
	bool
	empty(void) const { return false; }

	/**
		Linear-time size evaluation.  Always >= 1.
	 */
	size_type
	size(void) const { return std::distance(begin(), end()); }

	size_type
	max_size(void) const { return size_type(-1); }

	using base_type::contains;

	// front() ?
	// back() ?

	void
	push(const value_type& v);

	// don't allow use of iterators to insert...
	// only provide value interface.  

#if 0
	void
	insert(iterator pos, const value_type& v);
#endif

	/**
		Make sure that n is NOT already a member of this ring.  
		Invariant: this does not contain n if and only if n 
			does not contain this if and only if neither n 
			nor this contain any of each other's elements -- 
			they are disjoint.  
		This should be fairly type independent.
	 */
	void
	merge(this_type& n) {
		base_type::safe_merge(n);
	}

#if 0
	iterator
	erase(iterator pos);
#endif

	// void swap(...); ?

	// void clear(void); ?

	// void splice(...); ?

	// void remove(const value_type& v); ?

	// void sort(...); ?

};	// end class ring_node

//=============================================================================
}	// end namespace util

#endif	// __UTIL_RING_NODE_H__

