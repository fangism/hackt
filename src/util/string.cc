/**
	\file "util/string.cc"
	$Id: string.cc,v 1.6 2009/11/11 00:34:06 fang Exp $
	Additional string utilities.
 */

#include <string>
#include <cctype>
#include "util/string.tcc"
#include "util/macros.h"

namespace util {
namespace strings {
using std::string;

//=============================================================================
bool
string_begins_with(const string& s, const string& pre) {
	return s.find(pre) == 0;
}

//=============================================================================
/**
	Exact match substitution.
 */
size_t
strgsub(string& t, const string& s, const string& r) {
	// substitution is done by copying pieces of strings
	size_t subs = 0;
if (s.length() && (s != r)) {
	string ret;	// temporary
	const size_t slen = s.length();
	INVARIANT(slen);
	size_t tpos = 0;
	size_t mpos = t.find(s, tpos);
	while (mpos != string::npos) {
		++subs;
		// copy non-matching prefix
		ret.append(t.begin()+tpos, t.begin()+mpos);
		ret.append(r);		// substitute match
		tpos = mpos +slen;	// skip over matched part
		mpos = t.find(s, tpos);
	}
	ret.append(t.begin()+tpos, t.end());	// remainder
	if (t != ret)
		t = ret;
	// otherwise, just leave alone
}
	return subs;
}

//=============================================================================
/**
	Repeatedly removes prefix of string.
	\return the number of occurrences removed.  
 */
size_t
strip_prefix(string& t, const string& s) {
	size_t subs = 0;
	while (string_begins_with(t, s)) {
		t = t.substr(s.length());
		++subs;
	}
	return subs;
}

//=============================================================================
string
string_tolower(const string& s) {
	return transform_string(s, &tolower);
}

//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
string
string_toupper(const string& s) {
	return transform_string(s, &toupper);
}

//=============================================================================
}	// end namespace strings
}	// end namespace util

