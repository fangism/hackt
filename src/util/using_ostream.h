/**
	\file "using_ostream.h"
	Convenient using directives for ostream, cout, cerr, endl...
	Note: this is not wrapped by include-once protection, because it may
	be used within other namespaces.  
	Also assumes that <iosfwd> was included beforehand.  
	$Id: using_ostream.h,v 1.1 2004/12/05 05:07:26 fang Exp $
 */

#ifndef	__USING_OSTREAM_H__
#define	__USING_OSTREAM_H__

using std::ostream;
using std::cout;
using std::cerr;
using std::endl;
// any others?

#endif	// __USING_OSTREAM_H__
#undef	__USING_OSTREAM_H__


