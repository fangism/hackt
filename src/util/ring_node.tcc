/**
	\file "ring_node.tcc"
	Implementation of ring_node class.
	$Id: ring_node.tcc,v 1.1.2.2.6.3 2005/02/14 04:48:23 fang Exp $
 */

#ifndef	__UTIL_RING_NODE_TCC__
#define	__UTIL_RING_NODE_TCC__

#include "ring_node.h"

#if 0
#if defined(ENABLE_STACKTRACE) && ENABLE_STACKTRACE
	#include "stacktrace.h"
#else
	#define	STACKTRACE(x)
#endif
#endif

namespace util {
//=============================================================================

#if !FORCE_INLINE_RING_NODE
inline
ring_node_base::ring_node_base() : next(this) {
//	STACKTRACE("ring_node_base()");
#if 0
	std::cerr << "starting: " << this << " -> " << next << endl;
#endif
}

//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
inline
ring_node_base::ring_node_base(ring_node_base* r) : next(r) {
//	STACKTRACE("ring_node_base(*)");
	NEVER_NULL(next);
#if 0
	std::cerr << "starting: " << this << " -> " << next << endl;
#endif
}

//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
inline
ring_node_base::~ring_node_base() {
//	STACKTRACE("~ring_node_base()");
#if 0
	std::cerr << "end: " << this << " -> " << next << endl;
#endif
	if (next != this) {
		ring_node_base* walk = next;
		while (walk->next != this) {
#if 0
			std::cerr << "walking: " << walk << endl;
#endif
			walk = walk->next;
		}
		// found the node that points to this, update it
		walk->next = next;
	}
	// else this is the last node, just drops itself
	next = NULL;
}

//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
inline
void
ring_node_base::unsafe_merge(ring_node_base& r) {
//	STACKTRACE("ring_node_base::unsafe_merge()");
	std::swap(next, r.next);
}

//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
inline
bool
ring_node_base::contains(const ring_node_base& r) const {
	const_next_type walk1 = this;
	const_next_type walk2 = &r;
	do {
		if (walk1 == &r || walk2 == this)
			return true;
		else {
			walk1 = walk1->next;
			walk2 = walk2->next;
		}
	} while (walk1 != this && walk2 != &r);
	return false;
}

#endif	// FORCE_INLINE_RING_NODE

//=============================================================================
// class ring_node_derived method definitions

template <class T>
ring_node_derived<T>::~ring_node_derived() {
	if (next != this) {
		next_type walk = next;
		while (walk->next != this) {
			walk = walk->next;
		}
		// found the node that points to this, update it
		walk->next = next;
	}
	// else this is the last node, just drops itself
	next = NULL;
}

//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
template <class T>
bool
ring_node_derived<T>::contains(const this_type& r) const {
	const_next_type walk1 = this;
	const_next_type walk2 = &r;
	do {
		if (walk1 == &r || walk2 == this)
			return true;
		else {
			walk1 = walk1->next;
			walk2 = walk2->next;
		}
	} while (walk1 != this && walk2 != &r);
	return false;
}

//=============================================================================
}	// end namespace util

#endif	// __UTIL_RING_NODE_TCC__

