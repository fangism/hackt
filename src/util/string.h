/**
	\file "util/string.h"
	Configure-detected string library header.  
	For now, this is really reserved for C++.
	$Id: string.h,v 1.2 2006/01/22 06:53:37 fang Exp $
 */

#ifndef	__UTIL_STRING_H__
#define	__UTIL_STRING_H__

#include "config.h"
#include "util/NULL.h"
#include "util/libc.h"

#ifdef	__cplusplus
	#if	defined(HAVE_CSTRING) && HAVE_CSTRING
	#include <cstring>
	#endif
#else
	#if	defined(HAVE_STRING_H) && HAVE_STRING_H
	#include <string.h>
	#endif
	#if	defined(HAVE_STRINGS_H) && HAVE_STRINGS_H
	#include <strings.h>
	#endif
#endif

BEGIN_C_DECLS
// C-stuff here

END_C_DECLS

#ifdef	__cplusplus
namespace util {
namespace strings {

/**
	\param C is the character type, may be const, may be wchar_t.  
 */
template <class C>
inline
C*
eat_whitespace(C*& s) {
	if (s) {
		while (*s && isspace(*s))
			s++;
	}
	return s;
}


}	// end namespace strings
}	// end namespace util
#endif	/* __cplusplus */

#endif	/* __UTIL_STRING_H__ */

