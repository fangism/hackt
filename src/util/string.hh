/**
	\file "util/string.hh"
	Configure-detected string library header.  
	For now, this is really reserved for C++.
	$Id: string.h,v 1.7 2009/11/11 00:34:07 fang Exp $
 */

#ifndef	__UTIL_STRING_H__
#define	__UTIL_STRING_H__

#include "config.h"
#include "util/NULL.h"
#include "util/libc.h"

#ifdef	__cplusplus
	#if	defined(HAVE_CSTRING) && HAVE_CSTRING
	#include <cstring>
	#endif
	#include "util/string_fwd.hh"
#else
	#if	defined(HAVE_STRING_H) && HAVE_STRING_H
	#include <string.h>
	#endif
	#if	defined(HAVE_STRINGS_H) && HAVE_STRINGS_H
	#include <strings.h>
	#endif
#endif

BEGIN_C_DECLS
// C-stuff here

END_C_DECLS

#ifdef	__cplusplus
namespace util {
namespace strings {

//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
/**
	\param C is the character pointer/iterator type, 
	may be const, may be wchar_t.  
 */
template <class C>
inline
C
eat_whitespace(C& s) {
	if (s) {
		while (*s && isspace(*s))
			++s;
	}
	return s;
}

//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
/// Return false if conversion-assignment to int is successful.  
template <class I>
bool
string_to_num(const std::string&, I&);

//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
// regex-like functions
// waiting for std::regex <c++0x>

/**
	String global substitution.
	This is only based on exact-matching, not regular expressions!
	\param t the target string (modified)
	\param s the search pattern
	\param r the replacement string
	\return number of substitutions performed
 */
size_t
strgsub(std::string& t, const std::string& s, const std::string& r);


//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
size_t
strip_prefix(std::string& t, const std::string& s);

//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
bool
string_begins_with(const std::string&, const std::string&);

//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
template <typename Trans>
std::string
transform_string(const std::string&, Trans);

std::string
string_tolower(const std::string&);

std::string
string_toupper(const std::string&);

//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
}	// end namespace strings
}	// end namespace util
#endif	/* __cplusplus */

#endif	/* __UTIL_STRING_H__ */

