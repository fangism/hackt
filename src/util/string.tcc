/**
	\file "util/string.tcc"
	Implementations of some useful string functions.  
	$Id: string.tcc,v 1.6 2011/03/30 20:59:26 fang Exp $
 */

#ifndef	__UTIL_STRING_TCC__
#define	__UTIL_STRING_TCC__

#include <string>
#include <iterator>
#include <algorithm>
#include "util/string.hh"
#include "util/sstream.hh"

#include "util/numeric/sign_traits.hh"

namespace util {
namespace strings {
using std::string;
using std::stringstream;
using std::istringstream;
using std::ostringstream;
//=============================================================================
/**
	\param S true if type is signed
	\return true if rejected.
 */
template <bool S>
inline
bool
__reject_unsigned_negative(const char) {
	// negatives are allowed for signed, of course
	return false;
}

/**
	Specialization for unsigned, reject negatives.
 */
template <>
inline
bool
__reject_unsigned_negative<false>(const char f) {
	return f == '-';
}

//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
/**
	Helper function for converting string to int.  
	Just a wrapper around using stringstream to convert to int.  
	\param I some integer type.  
	\return true on error.  

	See: http://gcc.gnu.org/bugzilla/show_bug.cgi?id=39802
	In older versions, parsing -1 into an unsigned type
	would fail-to-fail.  
 */
template <class I>
bool
string_to_num(const string& s, I& i) {
	// prefer this over error-prone libc functions.  
	istringstream str(s);
	if (s.length() >= 2 && s[0] == '0' && s[1] == 'x') {
		str >> std::hex;
	}
	str >> i;
#if !defined(CXX_ISTREAM_NEGATIVE_UNSIGNED_FAILS)
	// workaround compiler/library bug
	if (s.length() && __reject_unsigned_negative<
			util::numeric::is_signed<I>::value>(s[0]))
		return true;
#endif
	return str.fail();
}

//- - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
/**
	Common routing for string transformations.
	\return new transformed copy of string.
 */
template <typename T>
string
transform_string(const string& s, T t) {
	string ret;
	std::transform(s.begin(), s.end(), std::back_inserter(ret), t);
	return ret;
}

//=============================================================================
}	// end namespace strings
}	// end namespace util

#endif	// __UTIL_STRING_TCC__

