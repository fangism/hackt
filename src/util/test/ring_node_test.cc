/**
	\file "ring_node_test.cc"
	Test for ring_node structures.  
	$Id: ring_node_test.cc,v 1.4 2006/02/26 02:28:05 fang Exp $
 */

#ifdef	NDEUBG
#undef	NDEBUG
#endif

#include <iostream>
#include <string>
#include "util/ring_node.tcc"

#include "util/using_ostream.hh"
using std::string;
using util::ring_node;
using util::ring_node_derived;

// typedef	ring_node<string>	ring_type;

template <class RingType>
ostream&
dump_ring_node(ostream& o, const RingType& r) {
	// could even use ostream_iterator
	typename RingType::const_iterator i = r.begin();
	const typename RingType::const_iterator e = r.end();
	for ( ; i!=e; i++)
		cout << *i << endl;
	return o;
}

template <template <class> class>
static
void
ring_type_test(void);

//=============================================================================
int
main(int, char*[]) {
	cout << "========= testing ring_node<string> =========" << endl;
	ring_type_test<ring_node>();
	cout << "===== testing ring_node_derived<string> =====" << endl;
	ring_type_test<ring_node_derived>();
	return 0;
}

//-----------------------------------------------------------------------------
template <template <class> class R>
static
void
ring_type_test(void) {
	typedef	R<string>	ring_type;
{
	ring_type first("Hello.");
	dump_ring_node(cout, first) << endl;
	assert(first.size() == 1);
}
{
	ring_type first("Hello.");
	ring_type second("Goodbye.");
	first.merge(second);
	assert(first.contains(second));
	assert(second.contains(first));
	assert(first.size() == 2);
	dump_ring_node(cout, first) << endl;
	dump_ring_node(cout, second) << endl;
}
{
	ring_type first("Hello.");
	ring_type second("Yoo-hoo.");
	{
		ring_type third("Goodbye.");
		first.merge(second);
		second.merge(third);
		assert(first.size() == 3);
		assert(first.contains(second));
		assert(first.contains(third));
		assert(second.contains(first));
		assert(second.contains(third));
		assert(third.contains(first));
		assert(third.contains(second));
		dump_ring_node(cout, first) << endl;
		dump_ring_node(cout, second) << endl;
		dump_ring_node(cout, third) << endl;
	}
	assert(first.size() == 2);
	dump_ring_node(cout, first) << endl;
	dump_ring_node(cout, second) << endl;
}
{
	ring_type first("Hello.");
	ring_type second("Yoo-hoo.");
	first.merge(second);
	assert(first.size() == 2);
	dump_ring_node(cout, first) << endl;
	dump_ring_node(cout, second) << endl;
{
	ring_type third("Durhhhh...");
	ring_type fourth("Goodbye.");
	third.merge(fourth);
	assert(third.size() == 2);
	dump_ring_node(cout, third) << endl;
	dump_ring_node(cout, fourth) << endl;
	second.merge(fourth);
	assert(first.contains(third));
	assert(first.contains(fourth));
	assert(second.contains(third));
	assert(second.contains(fourth));
	assert(third.contains(first));
	assert(third.contains(second));
	assert(fourth.contains(first));
	assert(fourth.contains(second));
	dump_ring_node(cout, first) << endl;
	dump_ring_node(cout, second) << endl;
	dump_ring_node(cout, third) << endl;
	dump_ring_node(cout, fourth) << endl;
	fourth.merge(first);	// already included, but safely handled
	dump_ring_node(cout, fourth) << endl;
}
	// after losing a few nodes, structure is still a ring!
	dump_ring_node(cout, first) << endl;
	dump_ring_node(cout, second) << endl;
}
}

