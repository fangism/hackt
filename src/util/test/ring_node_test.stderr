========= testing ring_node<string> =========
Hello.

Goodbye.
Hello.

Hello.
Goodbye.

Yoo-hoo.
Goodbye.
Hello.

Goodbye.
Hello.
Yoo-hoo.

Hello.
Yoo-hoo.
Goodbye.

Yoo-hoo.
Hello.

Hello.
Yoo-hoo.

Yoo-hoo.
Hello.

Hello.
Yoo-hoo.

Goodbye.
Durhhhh...

Durhhhh...
Goodbye.

Yoo-hoo.
Durhhhh...
Goodbye.
Hello.

Durhhhh...
Goodbye.
Hello.
Yoo-hoo.

Goodbye.
Hello.
Yoo-hoo.
Durhhhh...

Hello.
Yoo-hoo.
Durhhhh...
Goodbye.

Hello.
Yoo-hoo.
Durhhhh...
Goodbye.

Yoo-hoo.
Hello.

Hello.
Yoo-hoo.

===== testing ring_node_derived<string> =====
Hello.

Goodbye.
Hello.

Hello.
Goodbye.

Yoo-hoo.
Goodbye.
Hello.

Goodbye.
Hello.
Yoo-hoo.

Hello.
Yoo-hoo.
Goodbye.

Yoo-hoo.
Hello.

Hello.
Yoo-hoo.

Yoo-hoo.
Hello.

Hello.
Yoo-hoo.

Goodbye.
Durhhhh...

Durhhhh...
Goodbye.

Yoo-hoo.
Durhhhh...
Goodbye.
Hello.

Durhhhh...
Goodbye.
Hello.
Yoo-hoo.

Goodbye.
Hello.
Yoo-hoo.
Durhhhh...

Hello.
Yoo-hoo.
Durhhhh...
Goodbye.

Hello.
Yoo-hoo.
Durhhhh...
Goodbye.

Yoo-hoo.
Hello.

Hello.
Yoo-hoo.

